### PR TITLE
feat: Scene 4

### PR DESCRIPTION
**Pull request by ivanma9**
Scene 4 approved by Katherine, waiting for design team for any small fixes

### CODE REVIEW

File: src/components/pages/Story1/index.tsx [hunk 1]
```diff
@@ -3,18 +3,31 @@ import { useHistory } from 'react-router';
 
 import Scene1_fg from '../../../assets/scene1.png';
 import Scene2_fg from '../../../assets/scene2.png';
+
```
Comment (BryanPan342) [inline src/components/pages/Story1/index.tsx]: remove space

File: src/components/pages/Story1/index.tsx [hunk 2]
```diff
@@ -3,18 +3,31 @@ import { useHistory } from 'react-router';
...
 
   const content: SceneProps[] = [
+    {
+      sketch: Scene4_bg,
+      expo: [
+        {
+          text: 'blah blah blah more text.... asjdfalskdfjlqiwkejflidsf',
+          duration: 3000,
+        },
+      ],
```
Comment (BryanPan342) [inline src/components/pages/Story1/index.tsx]: plox fill in text

File: src/components/sketches/Scene4.ts [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    canvas = p.createCanvas(window.innerWidth, window.innerHeight);
+    canvas.id('p5-background');
+    p.pixelDensity(1);
+    cols = p.floor(window.innerWidth / SCL);
+    rows = p.floor(window.innerHeight / SCL);
+    p.background(0);
+
+    for (let i = 0; i < NUM_PARTICLES; i++) {
+      particles[i] = new Particle();
+    }
```
Comment (BryanPan342) [inline src/components/sketches/Scene4.ts]: nit: convert to map

File: src/components/sketches/Scene4.ts [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+    flowfield = new Array(rows * cols);
+
+  };
```
Comment (BryanPan342) [inline src/components/sketches/Scene4.ts]: remove weird spacing

File: src/components/sketches/Scene4.ts [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+  p.draw = () => {
+    if (showField) {
+      p.background(0);
+    } else {
+      p.background(p.color(0, 0, 0, 5));
+    }
```
Comment (BryanPan342) [inline src/components/sketches/Scene4.ts]:
```suggestion
    p.background(showField ? 0 : p.color(0, 0, 0, 5));
```

File: src/components/sketches/Scene4.ts [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+        for (let i = 0; i < rnd; i++) {
+          particles.shift();
+        }
+      }
+    }
+  };
+
+  p.windowResized = () => {
+    p.resizeCanvas(window.innerWidth, window.innerHeight, false);
+  };
```
Comment (BryanPan342) [inline src/components/sketches/Scene4.ts]: should this redraw on resize?

File: src/components/sketches/Scene4.ts [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+          p.stroke(255);
+          p.translate(x * SCL, y * SCL);
+          p.rotate(v.heading());
+          const endpoint = p.abs(m) * SCL;
+          p.line(0, 0, endpoint, 0);
+          if (m < 0) {
+            p.stroke('red');
+          } else {
+            p.stroke('green');
+          }
```
Comment (BryanPan342) [inline src/components/sketches/Scene4.ts]:
```suggestion
          p.stroke(m < 0 ? 'red' : 'green');
```

File: src/components/sketches/Scene4.ts [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+  };
+
+  p.draw = () => {
+    if (showField) {
+      p.background(0);
+    } else {
+      p.background(p.color(0, 0, 0, 5));
+    }
+    let yoff = start;
+    for (let y = 0; y < rows; y++) {
```
Comment (BryanPan342) [inline src/components/sketches/Scene4.ts]: i feel like this can be converted to a map but ill let you decide if you want to make it that

File: src/components/sketches/Scene4.ts [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+      let xoff = start;
+      for (let x = 0; x < cols; x++) {
```
Comment (BryanPan342) [inline src/components/sketches/Scene4.ts]:
```suggestion
      for (let x = 0; x < cols; x++ && xoff += INC) {
```

File: src/components/sketches/Scene4.ts [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+        flowfield[index] = v;
+        xoff += INC;
+      }
+      yoff += INC;
+    }
+    magOff += MAG_INC;
+    zoff += INC_START;
+    start -= MAG_INC;
+
+    if (!showField) {
```
Comment (BryanPan342) [inline src/components/sketches/Scene4.ts]: nit:

```
if (showField) return;
... rest of your implementation NOT in a nested If statement
```

Comment (BryanPan342) [reply src/components/sketches/Scene4.ts]: generally this is seen as more readable because you don't end up having a ton of nested loops

File: src/components/sketches/Scene4.ts [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+        for (let i = 0; i < rnd; i++) {
+          particles.push(new Particle());
+        }
+      } else if (particles.length > 2000) {
+        const rnd = p.floor(p.random(10));
+        for (let i = 0; i < rnd; i++) {
+          particles.shift();
+        }
+      }
+    }
```
Comment (BryanPan342) [inline src/components/sketches/Scene4.ts]: i think this can be combined in a cleaner more readable fashion!

File: src/components/sketches/Scene4.ts [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+  p.windowResized = () => {
+    p.resizeCanvas(window.innerWidth, window.innerHeight, false);
+  };
+
+  class Particle {
+    public pos: p5.Vector;
+    public vel: p5.Vector;
+    private _acc: p5.Vector;
+    private _maxSpeed: number;
+    public _prevPos: p5.Vector;
```
Comment (BryanPan342) [inline src/components/sketches/Scene4.ts]:
```suggestion
    private _prevPos: p5.Vector;
```

_ => private

File: src/components/sketches/Scene4.ts [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+      this.vel.limit(this._maxSpeed);
+      this.pos.add(this.vel);
+      this._acc.mult(0);
+    }
+
+    public show() {
+      p.strokeWeight(STROKE_WEIGHT);
+      p.line(this.pos.x, this.pos.y, this._prevPos.x, this._prevPos.y);
+      this.updatePrev();
+      //point(this.pos.x, this.pos.y);
```
Comment (BryanPan342) [inline src/components/sketches/Scene4.ts]: remove

File: src/components/sketches/Scene4.ts [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+        p.stroke(p.color(77, 184, 165));
+        if (this._bubble < BUBBLE_FREQUENCY) {
+          if (p.random() < BUBBLE_FREQUENCY_ON_LINES) {
+            p.fill(77, 184, 165);
+            p.ellipse(this.pos.x, this.pos.y, p.width*.006);
+          }
+        }
+      } else {
+        p.stroke(`rgba(${41}, ${157}, ${116}, ${TRANSPARENCY})`);
+      }
```
Comment (BryanPan342) [inline src/components/sketches/Scene4.ts]:
```suggestion
      p.stroke(di <= TRANSPARENCY_RADIUS ? p.color(77, 184, 165) : `rgba(${41}, ${157}, ${116}, ${TRANSPARENCY})`);
      if (this._bubble < BUBBLE_FREQUENCY && p.random() < BUBBLE_FREQUENCY_ON_LINES) {
        p.fill(77, 184, 165);
        p.ellipse(this.pos.x, this.pos.y, p.width*.006);
      }
```

File: src/components/sketches/Scene4.ts [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+    public edges() {
+      this.updatePrev();
+    }
+
+    public follow(vectors: p5.Vector[]) {
+      const x = p.floor(this.pos.x / SCL);
+      const y = p.floor(this.pos.y / SCL);
+      const index = x + y * cols;
```
Comment (BryanPan342) [inline src/components/sketches/Scene4.ts]: let's make this a helper function? i see it in the for loop up above

File: src/components/sketches/Scene4.ts [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+    public updatePrev() {
+      this._prevPos.x = this.pos.x;
+      this._prevPos.y = this.pos.y;
+    }
+
+    public edges() {
+      this.updatePrev();
+    }
```
Comment (BryanPan342) [inline src/components/sketches/Scene4.ts]: if this only servers the purpoose of doing `this.updatePrev()` lets just call `this.updatePrev()`

File: src/components/sketches/Scene4.ts [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+    public vel: p5.Vector;
+    private _acc: p5.Vector;
+    private _maxSpeed: number;
+    public _prevPos: p5.Vector;
+    private _bubble: number;
+
+    constructor() {
+      this.pos = p.createVector(p.random(window.innerWidth),p.random(window.innerHeight));
+      this.vel = p.createVector(0, 0);
+      this._acc = p.createVector(0, 0);
```
Comment (BryanPan342) [inline src/components/sketches/Scene4.ts:98]: remove all `_acc` they arent being used

**Review by BryanPan342 — CHANGES_REQUESTED** [review]
Looking good! Needs a little clean up but the logic is pretty sound

File: src/components/sketches/Scene4.ts (hunk 3 above)
Comment (ivanma9) [reply src/components/sketches/Scene4.ts]: `	
particles = Array(NUM_PARTICLES).fill(0).map(() => {
      return new Particle();
    });`

shouldnt this work? it's not..

Comment (BryanPan342) [reply src/components/sketches/Scene4.ts]: Particles needs to be declared with let not const

Comment (ivanma9) [reply src/components/sketches/Scene4.ts]: ive tried many different syntaxes to get map to work, and none of them are working even spread operator for a deep copy, fill, new, idk how to fix it

File: src/components/sketches/Scene4.ts (hunk 8 above)
Comment (ivanma9) [reply src/components/sketches/Scene4.ts]: maybe ill stab it another time. ik you love maps

File: src/components/sketches/Scene4.ts (hunk 10 above)
Comment (ivanma9) [reply src/components/sketches/Scene4.ts]: deleted showfield, not relevant for the code.

File: src/components/sketches/Scene4.ts (hunk 6 above)
Comment (ivanma9) [reply src/components/sketches/Scene4.ts]: Yes

File: src/components/sketches/Scene4.ts (hunk 17 above)
Comment (ivanma9) [reply src/components/sketches/Scene4.ts:98]: heh wdym. the `_vel `adds the `_acc`

File: src/components/sketches/Scene4.ts (hunk 16 above)
Comment (ivanma9) [reply src/components/sketches/Scene4.ts]: sounds good

File: src/components/sketches/Scene4.ts (hunk 15 above)
Comment (ivanma9) [reply src/components/sketches/Scene4.ts]: which part? `const index = x + y * cols;`?

Comment (ivanma9) [reply src/components/sketches/Scene4.ts]: I made a change on this, lmk if thats what you meant

File: src/components/sketches/Scene4.ts (hunk 14 above)
Comment (ivanma9) [reply src/components/sketches/Scene4.ts]: this suggestion makes bubbles every where, but we only want it on hover area within `di`

File: src/components/pages/Story1/index.tsx (hunk 2 above)
Comment (ivanma9) [reply src/components/pages/Story1/index.tsx]: @forrestburton got it B)

File: src/components/sketches/Scene4.ts (hunk 3 above)
Comment (BryanPan342) [reply src/components/sketches/Scene4.ts]: @ivanma9 

```
const particles: Particle[] = [];
```

this const on line 22 should be the problem as is said in the previous comment

change to:

```
let particles: Particle[] = [];
```

File: src/components/sketches/Scene4.ts (hunk 17 above)
Comment (BryanPan342) [reply src/components/sketches/Scene4.ts:98]: o my bad just read it

File: src/components/sketches/Scene4.ts (hunk 14 above)
Comment (BryanPan342) [reply src/components/sketches/Scene4.ts]: okay then just modify to make it work

```ts
      p.stroke(di <= TRANSPARENCY_RADIUS ? p.color(77, 184, 165) : `rgba(${41}, ${157}, ${116}, ${TRANSPARENCY})`);
      if (di <= TRANSPARENCY_RADIUS && this._bubble < BUBBLE_FREQUENCY && p.random() < BUBBLE_FREQUENCY_ON_LINES) {
        p.fill(77, 184, 165);
        p.ellipse(this.pos.x, this.pos.y, p.width*.006);
      }
```

File: src/components/sketches/Scene4.ts (hunk 3 above)
Comment (ivanma9) [reply src/components/sketches/Scene4.ts]: yeet i changed it back for linters just before recent push, but when I was testing out different map methods i used `let` instead of const.

Comment (ivanma9) [reply src/components/sketches/Scene4.ts]: basically, it still doesnt work with `let`

File: src/components/sketches/Scene4.ts (hunk 14 above)
Comment (ivanma9) [reply src/components/sketches/Scene4.ts]: yes that works out

File: src/utils/constants.ts [hunk 18]
```diff
@@ -40,4 +56,3 @@ export const EXPLODE_OFFSET = 15; // after a random number is generated from 1-E
 export const FIREWORK_LIFESPAN = 255;
 export const LIFESPAN_DECREMENT = 5;
 export const VELOCITY_CHANGE = 0.95;
-export const TRAIL_SIZE = 10;
```
Comment (BryanPan342) [inline src/utils/constants.ts:43]: why is this deleted?

Comment (ivanma9) [reply src/utils/constants.ts:43]: Yse, ty

Comment (ivanma9) [reply src/utils/constants.ts:43]: we need it

**Review by BryanPan342 — APPROVED** [review]
LGTM
